[ FEAT ] Add support for terminal variables

[Shell Parameter Expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) is a very useful feature available on operating systems terminals.   
They allow you to define system variables, so, it becomes easy to use terminal.   
For example, 
```sql
SELECT NAME FROM $HOME
```
It's much easier than write    
```sql
SELECT NAME FROM /home/my_user
```

We can get the environment variable from this [lib call](https://man7.org/linux/man-pages/man3/getenv.3.html).
In windows, we can have something like [this](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-wgetenv?view=msvc-160)

Tried to use `getenv` on Windows, but, Windwos says that this function is unsafe, and, should not be used ...
Compiler message: 
> D:\a\Termsequel\Termsequel\src\system.cpp(822,30): warning C4996: 'getenv': This function or variable may be unsafe. Consider using _dupenv_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. 

So, we maybe use the ["secure"](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-s-wgetenv-s?view=msvc-160) version of `getenv`, or, as the compiler [suggests](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/dupenv-s-wdupenv-s?view=msvc-160) ... 

This compiler message is a *warning*, not an error, but, since we threat warnings as errors, the compiler returns an error.

--- a/Changelog.md
+++ b/Changelog.md
@@ -63,4 +63,5 @@ Contains information about the versions of Termsequel
         | Support for negate expressions | [ISSUE](https://github.com/sgtcortez/Termsequel/issues/40) | [PR]() | Add three new operators, to be used as the negate form |
         | Add support for start pseudo column | [ISSUE](https://github.com/sgtcortez/Termsequel/issues/51) | [PR](https://github.com/sgtcortez/Termsequel/pull/57) | Add support for the **\***(STAR) pseudo column. |
         | Add RELATIVE_PATH column | [ISSUE](https://github.com/sgtcortez/Termsequel/issues/53) | [PR](https://github.com/sgtcortez/Termsequel/pull/60) | Added the relative path column |
-        | Add ABSOLUTE_PATH column | [ISSUE](https://github.com/sgtcortez/Termsequel/issues/52) | [PR](https://github.com/sgtcortez/Termsequel/pull/64) | Added the absolute path column |
+        | Add ABSOLUTE_PATH column | [ISSUE](https://github.com/sgtcortez/Termsequel/issues/52) | [PR](https://github.com/sgtcortez/Termsequel/pull/64) | Added the absolute path column |  
+        | Add support for variables | [ISSUE]() | [PR](https://github.com/sgtcortez/Termsequel/pull/69) | Now, it is possible to use environment variables, and terminal variables(if supported by your terminal) with termsequel **FROM** source.  

--- a/README.md
+++ b/README.md
@@ -11,6 +11,7 @@
         + [Columns](#Columns)
         + [SELECT](#SELECT)
         + [WHERE](#WHERE)
+        + [Variables](#Variables)
 - [Developers](#Developers)
 
 # Termsequel
@@ -189,6 +190,29 @@ $ termsequel 'SELECT NAME FROM /HOME'
         SELECT NAME FROM . WHERE NAME NOT_ENDS_WITH .txt
         ```
 
+    - ### Variables
+
+        Now, it is possible to use your environment variables with **Termsequel**!   
+
+        For example, in bash, you can write:    
+        ```shell
+        # This is the same to write: /home/my_user 
+        $ termsequel 'SELECT NAME FROM $HOME WHERE NAME ENDS_WITH .c OR NAME ENDS_WITH .cpp'
+        ```
+
+        If you terminal support, you even use non environment variables:   
+        **Example:** Exporting a user defined shell variable(will live only in this shel session)   
+        ```shell
+        $ export BOOKS=/home/sgtcortez/Documentos/Books/
+        $ termsequel 'SELECT NAME, SIZE FROM $BOOKS'
+        ```
+
+        **Note:** You can only use this feature at the first part!   
+        For example, `$HOME/$OTHER` will be expanded to: `/home/my_user/$OTHER`.  
+        There are no plans to support more than one ...     
+        
+        **Note:** The **$** character must be the first source character!
+        For example, `Documents$HOME` will not be expanded! 
 
 
 # Developers

--- a/include/system.hpp
+++ b/include/system.hpp
@@ -73,7 +73,7 @@ namespace Termsequel {
       public:
 
          // Executes the command, and return the list of elements
-         static void execute(const SystemCommand *command);
+         static bool execute(const SystemCommand *command);
    };
 
 }

--- a/src/system.cpp
+++ b/src/system.cpp
@@ -75,6 +75,9 @@ constexpr static const char FILE_SEPARATOR = '\\';
 
 #define BUFFER_SIZE 1024 * 4 // 4 KiB
 
+// Used to indicate a variable ...
+static constexpr char VARIABLE_SYMBOL = '$';
+
 #ifdef DEBUG
 #define DEBUG_SYSTEM 1
 #endif
@@ -152,6 +155,10 @@ static bool should_go_recursive(
    struct Termsequel::ConditionList *condition_list
 );
 
+/**
+ * Returns the value from a variable from operating system
+*/
+static std::string get_variable_value(std::string name);
 
 #ifdef __linux__
 
@@ -163,9 +170,20 @@ static std::string get_owner_name(uid_t owner);
 
 #endif
 
-void Termsequel::System::execute(const Termsequel::SystemCommand *command) {
+bool Termsequel::System::execute(const Termsequel::SystemCommand *command) {
 
-  const auto stat_array = get_information(command->target, command->conditions, 0);
+   std::string target;
+   if (command->target[0] == VARIABLE_SYMBOL) {
+      target = get_variable_value(command->target);
+      if (target.size() == 0) {
+         // the variable was not found, does nothing then ...
+         return false;
+      }
+   } else {
+      target = command->target;
+   }
+
+  const auto stat_array = get_information(target, command->conditions, 0);
 
   // output rule. Values must be padded
   // Starts with the default column name
@@ -346,6 +364,7 @@ void Termsequel::System::execute(const Termsequel::SystemCommand *command) {
       delete stat_element;
    }
    delete stat_array;
+   return true;
 }
 
 static std::vector<struct StatResult *> * get_information(
@@ -794,4 +813,38 @@ static bool should_go_recursive(
     }
   }
   return true;
+}
+
+static std::string get_variable_value(std::string name) {
+
+   // Ignores the $ symbol ...
+   const char *variable_name = name.c_str() + 1;
+#ifdef __linux__
+   // we should not free result, otherwise, we might get an undefined behavior
+   const char *result = std::getenv(variable_name);
+   if (result == nullptr) {
+      // invalid variable ...
+      std::cerr << "Value not found for: " << name << std::endl;
+      return "";
+   }
+
+   return result;
+#elif defined(_WIN32)
+   size_t required_size = 0;
+   getenv_s(&required_size, NULL, 0, variable_name);
+   if (required_size == 0) {
+      // invalid variable ...
+      std::cerr << "Value not found for: " << name << std::endl;
+      return "";
+   }
+   char *result = (char *) malloc(sizeof(char) * required_size + 1);
+   getenv_s(&required_size, result, required_size, variable_name);
+
+   // copy the value to the string
+   std::string value = result;
+
+   // frees the manually allocated string
+   free(result);
+   return value;
+#endif
 }

--- a/termsequel.cpp
+++ b/termsequel.cpp
@@ -30,6 +30,7 @@ int main (
 
    char *binary_name = argv[0];
    struct SystemCommand *command;
+   bool ok = false;
 
 #ifdef __linux__
 
@@ -53,7 +54,7 @@ int main (
       std::string sql = argv[argc - 1];
       Compiler compiler(sql);
       command = compiler.compile();
-      if (command) System::execute(command);
+      if (command) ok = System::execute(command);
       else return 1;
    }
 
@@ -69,9 +70,8 @@ int main (
       std::string sql = argv[argc - 1];
       Compiler compiler(sql);
       command = compiler.compile();
-      if (command) System::execute(command);
+      if (command) ok = System::execute(command);
       else return 1;
-
    }
 
 #endif
@@ -86,8 +86,8 @@ int main (
    // then, I decided for now, to manually delete
    delete command;
 
-
-   return 0;
+   int exit_status = ok ? 0 : 1;
+   return exit_status;
 }
 
 void show_help(

